TRIO112: replace one-line nursery with regular function call

jakkdl opened the pull request:
Final remaining issue without a PR :muscle: 

Code is fugly, will make a pass tomorrow, but I think the check is done.

Zac-HD commented on tests/trio112.py:
```diff
@@ -0,0 +1,35 @@
+with trio.open_nursery() as n:  # error in theory
+    n.start_soon(lambda n: n + 1)
+
+# safe?
+with trio.open_nursery() as n, open("") as o:
+    n.start(...)
```
Bit suspicious for #22-like reasons, but since violating this rule is more of a style problem than a correctness issue let's err on the side of less alarms.

jakkdl replied:
Thinking about this one more, is it actually that big of a style problem to have multiple `withitem`s in a single `with`? And if the open was placed in it's own `with` it'd clearly be a TRIO112 (or TRIO302/111).

I changed it to be an error, but trivial to change it back.

jakkdl commented:
rebased onto main, added support for multiple withitems, cleaned up code.
Also touched on some other stuff, partially related to this. Don't think worth bothering with a separate PR for it (will only be a minor conflict with #22) but can split if you want.

jakkdl commented on flake8_trio.py:
```diff
@@ -269,7 +259,7 @@ def visit_Yield(self, node: ast.Yield):
 
     # ---- 109 ----
-    def check_109(self, node: ast.AsyncFunctionDef):
+    def check_for_trio109(self, node: ast.AsyncFunctionDef):
```
realized function names were inconsistent, harmonized them

jakkdl commented on flake8_trio.py:
```diff
@@ -333,10 +353,30 @@ def has_exception(node: Optional[ast.expr]) -> str:
 
 
 class Visitor102(Flake8TrioVisitor):
+    class TrioScope:
```
`get_matching_call` (previously `get_trio_scope`) no longer constructs a `TrioScope`, since only `Visitor102` actually cared about it. So moved it to a subclass

Zac-HD replied:
IMO this isn't an affirmative reason *to* move it, so I'd keep the diff a bit smaller by leaving it where it was.  On the other hand, nor is there any real reason to move it *back*!

jakkdl replied:
haha. I liked subclassing it for people reading it later on not encountering it through a commit diff, fun tension

jakkdl commented on flake8_trio.py:
```diff
@@ -63,32 +75,21 @@ def __eq__(self, other: Any) -> bool:
-    def __str__(self):
-        # Not supporting other ways of importing trio, per TRIO106
-        return f"trio.{self.funcname}"
+def get_matching_call(
+    node: ast.AST, *names: str, base: str = "trio"
```
A bit obscuring to default to `"trio"` when no longer having it in the function name, but probs fine.

jakkdl commented on flake8_trio.py:
```diff
@@ -11,7 +11,18 @@
 import ast
 import tokenize
-from typing import Any, Dict, Iterable, List, NamedTuple, Optional, Set, Union
+from typing import (
+    Any,
+    Dict,
+    Iterable,
+    List,
+    NamedTuple,
+    Optional,
+    Set,
+    Tuple,
+    Union,
+    cast,
+)
```
damn I long for when 3.8 is EOL, a pain to have to add stuff up here all the time (should look for if there's any good neovim plugins that suggests and automatically adds imports)

Zac-HD commented on flake8_trio.py:
```diff
@@ -290,18 +280,48 @@ def visit_Import(self, node: ast.Import):
+            # check for trio.open_nursery
+            nursery = get_matching_call(item.context_expr, "open_nursery")
+
+            # isinstance(..., ast.Call) is done in get_matching_call
+            body_call = cast(ast.Call, node.body[0].value)
```
I'd use an `assert` rather than a `cast()` here.

jakkdl replied:
`assert` is actually not possible here, the isinstance call is done later inside `get_matching_call` (which if it inside there evaluates false propagates to the condition calling it being false).

the easiest alternative is a redundant `isinstance` in the condition

I suppose the cast is incorrect in that `body_call` isn't actually a `Call` *yet*, but is for brevity/convenience later.

Zac-HD reviewed (approved):
LGTM!
